core/src/main/java/io/grpc/internal/ClientCallImpl.java

```diff
@@ -164,6 +172,26 @@ public void start(final Listener<RespT> observer, Metadata headers) {
     checkNotNull(observer, "observer");
     checkNotNull(headers, "headers");
 
+    // Create the context
+
+    // Same time used to compute all timeouts, in order to avoid inaccuracies and negative
+    // timeouts.
+    final long now = System.nanoTime();
+    final Deadline callDeadline = callOptions.getDeadlineNanoTime() != null
+        ? Deadline.after(callOptions.getDeadlineNanoTime() - now, NANOSECONDS)
```

Comment:
It's fine to do in this PR, assuming it doesn't grow to be a thing unto itself (I was thinking it would be fairly small). Separate PR works fine as well.